redis/src/sentinel.rs

```diff
@@ -0,0 +1,558 @@
+
+#[cfg(feature = "aio")]
+async fn async_try_sentinel_replicas(
+    connection_info: &ConnectionInfo,
+    master_name: &str,
+) -> RedisResult<Vec<HashMap<String, String>>> {
+    let sentinel_client = Client::open(connection_info.clone())?;
+    let mut conn = sentinel_client.get_async_connection().await?;
+    let result: Vec<HashMap<String, String>> = sentinel_replicas_cmd(master_name)
+        .query_async(&mut conn)
+        .await?;
+    Ok(result)
+}
+
+fn try_sentinel_masters(
+    connection_info: &ConnectionInfo,
+) -> RedisResult<Vec<HashMap<String, String>>> {
+    let sentinel_client = Client::open(connection_info.clone())?;
+    let mut conn = sentinel_client.get_connection()?;
+    let result: Vec<HashMap<String, String>> = sentinel_masters_cmd().query(&mut conn)?;
+    Ok(result)
+}
+
+fn try_sentinel_replicas(
+    connection_info: &ConnectionInfo,
+    master_name: &str,
+) -> RedisResult<Vec<HashMap<String, String>>> {
+    let sentinel_client = Client::open(connection_info.clone())?;
+    let mut conn = sentinel_client.get_connection()?;
+    let result: Vec<HashMap<String, String>> =
+        sentinel_replicas_cmd(master_name).query(&mut conn)?;
+    Ok(result)
+}
+
+fn is_master_valid(master_info: &HashMap<String, String>, service_name: &str) -> bool {
+    master_info.get("name").map(|s| s.as_str()) == Some(service_name)
+        && master_info.contains_key("ip")
+        && master_info.contains_key("port")
+        && master_info.get("flags").map_or(false, |flags| {
+            flags.contains("master") && !flags.contains("s_down") && !flags.contains("o_down")
+        })
+        && master_info["port"].parse::<u16>().is_ok()
+}
+
+fn is_replica_valid(replica_info: &HashMap<String, String>) -> bool {
+    replica_info.contains_key("ip")
+        && replica_info.contains_key("port")
+        && replica_info.get("flags").map_or(false, |flags| {
+            !flags.contains("s_down") && !flags.contains("o_down")
+        })
+        && replica_info["port"].parse::<u16>().is_ok()
+}
+
+fn random_replica_index(max: usize) -> usize {
+    if max == 0 {
+        0
+    } else {
+        let nanos = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .subsec_nanos();
+        (nanos as usize) % max
+    }
+}
+
+fn try_connect_to_first_replica(
+    addresses: &Vec<(String, u16)>,
+    start_index: usize,
+) -> Result<Client, crate::RedisError> {
+    if addresses.is_empty() {
+        fail!((
+            ErrorKind::NoValidReplicasFound,
+            "No valid replica found in sentinel for given name",
+        ));
+    }
+
+    let mut last_err = None;
+    for i in 0..addresses.len() {
+        let index = (i + start_index) % addresses.len();
+        match Client::open(addresses[index].clone()) {
+            Ok(client) => return Ok(client),
+            Err(err) => last_err = Some(err),
+        }
+    }
+
+    // We can unwrap here because we know there is at least one address.
```

Comment:
I'll try to improve this comment further. My point was that if we have at least one address, we can guarantee there will be an error there (since any address would either store an error or just return a client).